conbench/entities/history.py

```diff
@@ -342,112 +343,89 @@ def _add_rolling_stats_columns_to_history_query(
     exclusive on the right side. This is useful if you want to compare each commit to
     the previous commit's rolling stats.
     """
-    # Define window ranges based on the server configuration
-    range_excluding_current_commit = (-Config.DISTRIBUTION_COMMITS, -1)
-    range_including_current_commit = (-Config.DISTRIBUTION_COMMITS + 1, 0)
-
-    # Clean up a few columns so they're more usable
-    query = Session.query(
-        query,
-        # Clean up begins_distribution_change so it's a non-null boolean column
-        s.func.coalesce(
-            query.c.change_annotations["begins_distribution_change"].as_boolean(),
-            False,
-        ).label("begins_distribution_change"),
-        # When we do rolling calculations, we want to roll over commits (not individual
-        # BenchmarkResults, since there may be multiple of those on one commit). Create
-        # a column that represents the commit order so we can roll over that.
-        s.func.dense_rank().over(order_by=query.c.timestamp).label("commit_rank"),
-    ).subquery()
+    # pandas likes the data to be sorted
+    df.sort_values(
+        ["case_id", "context_id", "hash", "repository", "timestamp"], inplace=True
+    )
 
-    # Add column with cumulative sum of distribution changes, to identify the segment
-    query = Session.query(
-        query,
-        s.func.sum(query.c.begins_distribution_change.cast(s.Integer))
-        .over(
-            partition_by=(
-                query.c.case_id,
-                query.c.context_id,
-                query.c.hash,
-                query.c.repository,
-            ),
-            order_by=query.c.commit_rank,
+    # Clean up begins_distribution_change so it's a non-null boolean column
+    df["begins_distribution_change"] = [
+        x.get("begins_distribution_change", False) for x in df["change_annotations"]
+    ]
+
+    # When we do rolling calculations, we want to roll over commits (not individual
+    # BenchmarkResults, since there may be multiple of those on one commit). Create a
+    # column that represents the commit order so we can roll over that.
+    df["commit_rank"] = (
+        df.groupby(["case_id", "context_id", "hash", "repository"])["timestamp"].rank(
+            method="dense"
         )
-        .label("segment_id"),
-    ).subquery()
+        # HACK: pandas needs a datetime-like column if you want windows with
+        # potentially-different numbers of data points.
+        # TODO: use a subclass from pd.api.indexers.BaseIndexer instead
+        .astype("datetime64[s]")
+    )
+
+    # Add column with cumulative sum of distribution changes, to identify the segment
+    df["segment_id"] = (
+        df.groupby(["case_id", "context_id", "hash", "repository"])
+        .rolling(window=f"{Config.DISTRIBUTION_COMMITS}s", on="commit_rank")[
```

Comment:
Shouldn't you just use the int as the window so it goes by observation instead of relying on a sequential 1s index since it's not time?

(Though to be honest, this exact parameter is one of the parts of the pandas API I've complained about the most; it's not just a union-typed param, it actually returns different classes with different methods depending on the type.)

Reply:
Yeah this API is really bad. The problem (as before) is we might have multiple observations per commit. So let's say `DISTRIBUTION_COMMITS=2`. I want something like this:
```
 ix | timestamp  | commit_rank | mean | rolled_ixs 
----+------------+-------------+------+------------
 0  | 2020-01-01 | 1           | 10   | -          
 1  | 2020-01-02 | 2           | 11   | 0          
 2  | 2020-01-02 | 2           | 12   | 0          
 3  | 2020-01-05 | 3           | 13   | 0,1,2      
 4  | 2020-01-07 | 4           | 14   | 1,2,3      
 5  | 2020-01-07 | 4           | 15   | 1,2,3      
```
Looking at the last row, the indices we want to use for `rolling_mean_excluding_this_commit` are 1,2,3. That's because we're on the Jan 7 commit, and we want to use data from the previous 2 commits (Jan 5 and Jan 2). That corresponds to indxes 1,2,3.

I was able to do this in SQL with a window function over `commit_rank` with RANGE BETWEEN 2 PRECEDING AND 1 PRECEDING. 

In pandas it's tricky. I can't do `window=2` because that assumes it's a constant 2 rows per window; i.e. the `rolled_ixs` for the last row would be 3,4. So I tried this hack. I'm not even sure it's working as intended; with this hack, the last two rows are returning different `rolled_ixs`.

Reply:
This is solved in the latest commit using [custom window rolling](https://pandas.pydata.org/docs/user_guide/window.html#custom-window-rolling).